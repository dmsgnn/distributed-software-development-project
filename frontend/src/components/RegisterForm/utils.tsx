import { post } from '@/utils/network';

export enum RegisterFields {
	EMAIL = 'register-email',
	FIRST_NAME = 'register-firstname',
	LAST_NAME = 'register-lastname',
	PASSWORD = 'register-password',
	PASSWORD_REPEAT = 'register-passwordrepeat',
}

export type RegisterFormValues = Record<RegisterFields, unknown>;

export function registerUser(formData: RegisterFormValues) {
	const requestBody = {
		firstName: formData[RegisterFields.FIRST_NAME],
		lastName: formData[RegisterFields.LAST_NAME],
		email: formData[RegisterFields.EMAIL],
		password: formData[RegisterFields.PASSWORD],
		secondPassword: formData[RegisterFields.PASSWORD_REPEAT],
	};

	if (requestBody.password !== requestBody.secondPassword) {
		throw new Error('Passwords do not match!');
	}

	return post('/api/auth/register', requestBody);
}
