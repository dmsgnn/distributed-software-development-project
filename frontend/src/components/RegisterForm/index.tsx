import { RegisterFields, RegisterFormValues, registerUser } from '@/components/RegisterForm/utils';
import { Button, FormControl, FormErrorMessage, FormLabel, Input, Text, useToast, VStack } from '@chakra-ui/react';
import { useRouter } from 'next/router';
import { useForm } from 'react-hook-form';
import useMutation from 'use-mutation';

export default function RegisterForm() {
	const {
		handleSubmit,
		register,
		formState: { errors, isSubmitting },
		setError,
	} = useForm<RegisterFormValues>({
		criteriaMode: 'all',
		mode: 'onBlur',
	});

	const toast = useToast();
	const router = useRouter();

	const [onSubmit] = useMutation(registerUser, {
		onSuccess: () => {
			toast({
				title: 'Successful registration',
				status: 'success',
				description: 'You have successfully registered your account',
			});
			router.push('/login');
		},
		onFailure: ({ error }) => {
			if ('fieldErrors' in error) {
				(error.fieldErrors as Array<{ fieldValue: string; message: string }>).forEach((fieldError) => {
					setError(`register-${fieldError.fieldValue}` as RegisterFields, { message: fieldError.message });
				});
			} else {
				toast({ title: 'Ops! Something went wrong', status: 'error', description: error?.message });
			}
		},
	});

	return (
		<form onSubmit={handleSubmit(onSubmit)}>
			<VStack align="stretch">
				<FormControl isInvalid={Boolean(errors[RegisterFields.EMAIL])}>
					<FormLabel>Email</FormLabel>
					<Input
						type="email"
						placeholder="Email"
						{...register(RegisterFields.EMAIL, {
							required: 'Email is required.',
							pattern: {
								value: /.*@.*/,
								message: 'Email is not valid.',
							},
						})}
					/>
					<FormErrorMessage>{errors[RegisterFields.EMAIL]?.message}</FormErrorMessage>
				</FormControl>

				<FormControl isInvalid={Boolean(errors[RegisterFields.FIRST_NAME])}>
					<FormLabel>First name</FormLabel>
					<Input
						placeholder="First name"
						{...register(RegisterFields.FIRST_NAME, {
							required: 'First name is required.',
							pattern: {
								value: /[\p{Letter}\p{Mark}]+/gu,
								message: 'First name must contain only letters.',
							},
							maxLength: {
								value: 32,
								message: 'First name should be only 32 characters long.',
							},
						})}
					/>
					<FormErrorMessage>{errors[RegisterFields.FIRST_NAME]?.message}</FormErrorMessage>
				</FormControl>

				<FormControl isInvalid={Boolean(errors[RegisterFields.LAST_NAME])}>
					<FormLabel>Last name</FormLabel>
					<Input
						placeholder="Last name"
						{...register(RegisterFields.LAST_NAME, {
							required: 'Last name is required.',
							pattern: {
								value: /[\p{Letter}\p{Mark}]+/gu,
								message: 'Last name must contain only letters.',
							},
							maxLength: {
								value: 32,
								message: 'Last name should be only 32 characters long.',
							},
						})}
					/>
					<FormErrorMessage>{errors[RegisterFields.LAST_NAME]?.message}</FormErrorMessage>
				</FormControl>

				<FormControl isInvalid={Boolean(errors[RegisterFields.PASSWORD])}>
					<FormLabel>Password</FormLabel>
					<Input
						type="password"
						placeholder="Password"
						{...register(RegisterFields.PASSWORD, {
							required: 'Password is required.',
							pattern: {
								value: /^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#$%^&*_=+-]).{10,}$/,
								message:
									'Password is too simple. Length must be 10 or more characters. Use uppercase, digit and special characters.',
							},
						})}
					/>
					<FormErrorMessage>{errors[RegisterFields.PASSWORD]?.message}</FormErrorMessage>
				</FormControl>

				<FormControl isInvalid={Boolean(errors[RegisterFields.PASSWORD_REPEAT])}>
					<FormLabel>Repeat password</FormLabel>
					<Input
						type="password"
						placeholder="Repeat password"
						{...register(RegisterFields.PASSWORD_REPEAT, {
							required: 'You have to repeat your password.',
							pattern: {
								value: /^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#$%^&*_=+-]).{10,}$/,
								message:
									'Password is too simple. Length must be 10 or more characters. Use uppercase, digit and special characters.',
							},
						})}
					/>
					<FormErrorMessage>{errors[RegisterFields.PASSWORD_REPEAT]?.message}</FormErrorMessage>
				</FormControl>

				<Button type="submit" colorScheme="blue" isLoading={isSubmitting} isDisabled={isSubmitting}>
					Sign up
				</Button>

				<Text textAlign="center">
					By creating an account, you are accepting our terms of service and privacy and cookie policy.
				</Text>
			</VStack>
		</form>
	);
}
