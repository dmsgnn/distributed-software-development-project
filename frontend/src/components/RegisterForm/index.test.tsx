import userEvent from '@testing-library/user-event';
import { axe } from 'jest-axe';

import { act, render, screen, waitFor } from 'test-utils';
import RegisterForm from './';

const mockPost = jest.fn();
jest.mock('@/utils/network', () => ({ post: (...args) => mockPost(...args) }));

const mockRouterPush = jest.fn();
jest.mock('next/router', () => ({ useRouter: jest.fn(() => ({ push: mockRouterPush })) }));

describe('RegisterForm', () => {
	it('should matches snapshot', () => {
		const { asFragment } = render(<RegisterForm />);

		expect(asFragment()).toMatchSnapshot();
	});

	it('should be accessible', async () => {
		const { container } = render(<RegisterForm />);

		let results: Awaited<ReturnType<typeof axe>>;
		await act(async () => {
			results = await axe(container);
		});

		expect(results).toHaveNoViolations();
	});

	it('should create an api call on submit (reject)', async () => {
		mockPost.mockResolvedValue({});
		const user = userEvent.setup();
		render(<RegisterForm />);

		const submitButton = screen.getByRole('button', { name: 'Sign up' });
		expect(submitButton).toBeInTheDocument();

		await user.type(screen.getByPlaceholderText('Email'), 'mail@mail.com');
		await user.type(screen.getByPlaceholderText('First name'), 'first name');
		await user.type(screen.getByPlaceholderText('Last name'), 'last name');
		await user.type(screen.getByPlaceholderText('Password'), 'Password123*');
		await user.type(screen.getByPlaceholderText('Repeat password'), 'Password123*');

		await act(async () => {
			await user.click(submitButton);
		});

		waitFor(() => expect(mockPost).toHaveBeenCalled());
	});

	it('should create an api call on submit (reject)', async () => {
		mockPost.mockRejectedValue({});
		const user = userEvent.setup();
		render(<RegisterForm />);

		const submitButton = screen.getByRole('button', { name: 'Sign up' });
		expect(submitButton).toBeInTheDocument();

		await user.type(screen.getByPlaceholderText('Email'), 'mail@mail.com');
		await user.type(screen.getByPlaceholderText('First name'), 'first name');
		await user.type(screen.getByPlaceholderText('Last name'), 'last name');
		await user.type(screen.getByPlaceholderText('Password'), 'Password123*');
		await user.type(screen.getByPlaceholderText('Repeat password'), 'Password123*');

		await act(async () => {
			await user.click(submitButton);
		});

		waitFor(() => expect(mockPost).toHaveBeenCalled());
	});

	it('should create an api call on submit (reject with fieldErrors)', async () => {
		mockPost.mockRejectedValue({ fieldErrors: [{ fieldValue: 'string', message: 'string' }] });
		const user = userEvent.setup();
		render(<RegisterForm />);

		const submitButton = screen.getByRole('button', { name: 'Sign up' });
		expect(submitButton).toBeInTheDocument();

		await user.type(screen.getByPlaceholderText('Email'), 'mail@mail.com');
		await user.type(screen.getByPlaceholderText('First name'), 'first name');
		await user.type(screen.getByPlaceholderText('Last name'), 'last name');
		await user.type(screen.getByPlaceholderText('Password'), 'Password123*');
		await user.type(screen.getByPlaceholderText('Repeat password'), 'Password123*');

		await act(async () => {
			await user.click(submitButton);
		});

		waitFor(() => expect(mockPost).toHaveBeenCalled());
	});
});
