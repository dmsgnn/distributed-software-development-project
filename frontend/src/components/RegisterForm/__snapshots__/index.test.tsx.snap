// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm should matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <div
      class="chakra-stack css-o5l3sd"
    >
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-uvho5g"
          for="field-:r0:"
          id="field-:r0:-label"
        >
          Email
        </label>
        <input
          class="chakra-input css-0"
          id="field-:r0:"
          name="register-email"
          placeholder="Email"
          type="email"
        />
      </div>
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-uvho5g"
          for="field-:r1:"
          id="field-:r1:-label"
        >
          First name
        </label>
        <input
          class="chakra-input css-0"
          id="field-:r1:"
          name="register-firstname"
          placeholder="First name"
        />
      </div>
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-uvho5g"
          for="field-:r2:"
          id="field-:r2:-label"
        >
          Last name
        </label>
        <input
          class="chakra-input css-0"
          id="field-:r2:"
          name="register-lastname"
          placeholder="Last name"
        />
      </div>
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-uvho5g"
          for="field-:r3:"
          id="field-:r3:-label"
        >
          Password
        </label>
        <input
          class="chakra-input css-0"
          id="field-:r3:"
          name="register-password"
          placeholder="Password"
          type="password"
        />
      </div>
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-uvho5g"
          for="field-:r4:"
          id="field-:r4:-label"
        >
          Repeat password
        </label>
        <input
          class="chakra-input css-0"
          id="field-:r4:"
          name="register-passwordrepeat"
          placeholder="Repeat password"
          type="password"
        />
      </div>
      <button
        class="chakra-button css-4xx2wk"
        type="submit"
      >
        Sign up
      </button>
      <p
        class="chakra-text css-xi606m"
      >
        By creating an account, you are accepting our terms of service and privacy and cookie policy.
      </p>
    </div>
  </form>
</DocumentFragment>
`;
